### PR TITLE
ci: replace PR status checks with unified ci.yml and scope deploy to main

### LINKED ISSUE #156: Update the github/workflows to allows run
<html><head></head><body><div class="markdown-document"><p><span data-markdown-start-index="0">This edge case happens because GitHub branch protection / status checks look for specific required check names (e.g., </span><code class="inline-code"><span data-markdown-start-index="119">api-test</span></code><span data-markdown-start-index="128"> or </span><code class="inline-code"><span data-markdown-start-index="133">web-test</span></code><span data-markdown-start-index="142">). When a workflow is skipped via top-level </span><code class="inline-code"><span data-markdown-start-index="187">paths:</span></code><span data-markdown-start-index="194"> filters, those status checks are never reported at all, causing pull requests to wait indefinitely.</span></p><p><span data-markdown-start-index="295">To circumvent this, </span><strong><span data-markdown-start-index="317">move path filtering inside the workflow using conditional job execution and create a mandatory aggregator check.</span></strong></p><hr><h3><span data-markdown-start-index="440">Recommended Solution: Changes Detection with Conditional Execution</span></h3><p><span data-markdown-start-index="507">Run the workflow on every PR, check which directories changed using </span><code class="inline-code"><span data-markdown-start-index="576">dorny/paths-filter</span></code><span data-markdown-start-index="595">, execute tests only for modified components, and report a universal passing status.</span></p><ucs-code-block code="name: CI

<code>
on:
  pull_request:
    branches: [main]

jobs:
  changes:
    runs-on: ubuntu-latest
    outputs:
      api: ${{ steps.filter.outputs.api }}
      web: ${{ steps.filter.outputs.web }}
    steps:
      - uses: actions/checkout@v4
      - uses: dorny/paths-filter@v3
        id: filter
        with:
          filters: |
            api:
              - 'src/api/**'
            web:
              - 'src/web/**'

  test-api:
    needs: changes
    if: ${{ needs.changes.outputs.api == 'true' }}
    runs-on: ubuntu-latest
    steps:
      - uses: actions/checkout@v4
      - name: Run API Tests
        run: npm test --prefix src/api

  test-web:
    needs: changes
    if: ${{ needs.changes.outputs.web == 'true' }}
    runs-on: ubuntu-latest
    steps:
      - actions/checkout@v4
      - name: Run Web Tests
        run: npm test --prefix src/web

  # Set this single job as your required status check in GitHub branch rules
  ci-check:
    runs-on: ubuntu-latest
    needs: [changes, test-api, test-web]
    if: always()
    steps:
      - name: Check Statuses
        run: |
          if [ &quot;${{ contains(needs.*.result, 'failure') }}&quot; = &quot;true&quot; ] || \
             [ &quot;${{ contains(needs.*.result, 'cancelled') }}&quot; = &quot;true&quot; ]; then
            echo &quot;A required test failed or was cancelled.&quot;
            exit 1
          fi
<code/>
          echo &quot;All checks passed or were safely skipped.&quot;
" language="yaml" spk2="" dark-theme="" next-gen="" luminous="" jslog="340197;track:impression;xFZKLc:W251bGwsInVjcy1jb2RlLWJsb2NrIixudWxsLG51bGwsbnVsbCxudWxsLG51bGwsbnVsbCxudWxsLG51bGwsMjJd"><pre><code class="language-yaml"><span>name: CI

on:
  pull_request:
    branches: [main]

</span><span class="hljs-built_in"><span>jobs</span></span><span>:
  changes:
    runs-on: ubuntu-latest
    outputs:
      api: </span><span class="hljs-variable"><span>${{ steps.filter.outputs.api }</span></span><span>}
      web: </span><span class="hljs-variable"><span>${{ steps.filter.outputs.web }</span></span><span>}
    steps:
      - uses: actions/checkout@v4
      - uses: dorny/paths-filter@v3
        id: filter
        with:
          filters: |
            api:
              - </span><span class="hljs-string"><span>'src/api/**'</span></span><span>
            web:
              - </span><span class="hljs-string"><span>'src/web/**'</span></span><span>

  test-api:
    needs: changes
    </span><span class="hljs-keyword"><span>if</span></span><span>: </span><span class="hljs-variable"><span>${{ needs.changes.outputs.api == 'true' }</span></span><span>}
    runs-on: ubuntu-latest
    steps:
      - uses: actions/checkout@v4
      - name: Run API Tests
        run: npm </span><span class="hljs-built_in"><span>test</span></span><span> --prefix src/api

  test-web:
    needs: changes
    </span><span class="hljs-keyword"><span>if</span></span><span>: </span><span class="hljs-variable"><span>${{ needs.changes.outputs.web == 'true' }</span></span><span>}
    runs-on: ubuntu-latest
    steps:
      - actions/checkout@v4
      - name: Run Web Tests
        run: npm </span><span class="hljs-built_in"><span>test</span></span><span> --prefix src/web

  </span><span class="hljs-comment"><span># Set this single job as your required status check in GitHub branch rules</span></span><span>
  ci-check:
    runs-on: ubuntu-latest
    needs: [changes, test-api, test-web]
    </span><span class="hljs-keyword"><span>if</span></span><span>: always()
    steps:
      - name: Check Statuses
        run: |
          </span><span class="hljs-keyword"><span>if</span></span><span> [ </span><span class="hljs-string"><span>"</span><span class="hljs-variable"><span>${{ contains(needs.*.result, 'failure') }</span></span><span>}"</span></span><span> = </span><span class="hljs-string"><span>"true"</span></span><span> ] || \
             [ </span><span class="hljs-string"><span>"</span><span class="hljs-variable"><span>${{ contains(needs.*.result, 'cancelled') }</span></span><span>}"</span></span><span> = </span><span class="hljs-string"><span>"true"</span></span><span> ]; </span><span class="hljs-keyword"><span>then</span></span><span>
            </span><span class="hljs-built_in"><span>echo</span></span><span> </span><span class="hljs-string"><span>"A required test failed or was cancelled."</span></span><span>
            </span><span class="hljs-built_in"><span>exit</span></span><span> 1
          </span><span class="hljs-keyword"><span>fi</span></span><span>
          </span><span class="hljs-built_in"><span>echo</span></span><span> </span><span class="hljs-string"><span>"All checks passed or were safely skipped."</span></span><span>
</span></code></pre></ucs-code-block><hr><h3><span data-markdown-start-index="2045">Alternative Approaches</span></h3><ucs-markdown-table encodedtablecmarknode="[24,null,null,null,null,null,null,null,null,null,[[26,null,null,null,null,null,null,null,null,null,[[25,null,null,null,null,null,null,null,null,null,[[12,null,null,null,&quot;Approach&quot;,null,null,null,null,null,null,null,null,null,[73,3,73,10]]],null,null,null,[73,2,73,11]],[25,null,null,null,null,null,null,null,null,null,[[12,null,null,null,&quot;How it works&quot;,null,null,null,null,null,null,null,null,null,[73,14,73,25]]],null,null,null,[73,13,73,26]],[25,null,null,null,null,null,null,null,null,null,[[12,null,null,null,&quot;Pros&quot;,null,null,null,null,null,null,null,null,null,[73,29,73,32]]],null,null,null,[73,28,73,33]],[25,null,null,null,null,null,null,null,null,null,[[12,null,null,null,&quot;Cons&quot;,null,null,null,null,null,null,null,null,null,[73,36,73,39]]],null,null,null,[73,35,73,40]]],null,null,null,[73,1,73,41],null,null,null,null,null,[1]],[26,null,null,null,null,null,null,null,null,null,[[25,null,null,null,null,null,null,null,null,null,[[19,null,null,null,null,null,null,null,null,null,[[12,null,null,null,&quot;Aggregator Job Pattern (Recommended)&quot;,null,null,null,null,null,null,null,null,null,[75,5,75,40]]],null,null,null,[75,3,75,42]]],null,null,null,[75,2,75,43]],[25,null,null,null,null,null,null,null,null,null,[[12,null,null,null,&quot;Uses &quot;,null,null,null,null,null,null,null,null,null,[75,46,75,50]],[15,null,null,null,&quot;paths-filter&quot;,null,null,null,null,null,null,null,null,null,[75,52,75,63]],[12,null,null,null,&quot; + a single required &quot;,null,null,null,null,null,null,null,null,null,[75,65,75,85]],[15,null,null,null,&quot;ci-check&quot;,null,null,null,null,null,null,null,null,null,[75,87,75,94]],[12,null,null,null,&quot; job evaluated with &quot;,null,null,null,null,null,null,null,null,null,[75,96,75,115]],[15,null,null,null,&quot;if: always()&quot;,null,null,null,null,null,null,null,null,null,[75,117,75,128]],[12,null,null,null,&quot;.&quot;,null,null,null,null,null,null,null,null,null,[75,130,75,130]]],null,null,null,[75,45,75,131]],[25,null,null,null,null,null,null,null,null,null,[[12,null,null,null,&quot;Scalable, clean GitHub branch protection setup, tests execute only when needed.&quot;,null,null,null,null,null,null,null,null,null,[75,134,75,212]]],null,null,null,[75,133,75,213]],[25,null,null,null,null,null,null,null,null,null,[[12,null,null,null,&quot;Requires defining the aggregator job.&quot;,null,null,null,null,null,null,null,null,null,[75,216,75,252]]],null,null,null,[75,215,75,253]]],null,null,null,[75,1,75,254],null,null,null,null,null,[0]],[26,null,null,null,null,null,null,null,null,null,[[25,null,null,null,null,null,null,null,null,null,[[19,null,null,null,null,null,null,null,null,null,[[12,null,null,null,&quot;Root Path Invalidation&quot;,null,null,null,null,null,null,null,null,null,[76,5,76,26]]],null,null,null,[76,3,76,28]]],null,null,null,[76,2,76,29]],[25,null,null,null,null,null,null,null,null,null,[[12,null,null,null,&quot;Add root config paths (e.g., &quot;,null,null,null,null,null,null,null,null,null,[76,32,76,60]],[15,null,null,null,&quot;package.json&quot;,null,null,null,null,null,null,null,null,null,[76,62,76,73]],[12,null,null,null,&quot;, &quot;,null,null,null,null,null,null,null,null,null,[76,75,76,76]],[15,null,null,null,&quot;firebase.json&quot;,null,null,null,null,null,null,null,null,null,[76,78,76,90]],[12,null,null,null,&quot;) to both &quot;,null,null,null,null,null,null,null,null,null,[76,92,76,101]],[15,null,null,null,&quot;api&quot;,null,null,null,null,null,null,null,null,null,[76,103,76,105]],[12,null,null,null,&quot; and &quot;,null,null,null,null,null,null,null,null,null,[76,107,76,111]],[15,null,null,null,&quot;web&quot;,null,null,null,null,null,null,null,null,null,[76,113,76,115]],[12,null,null,null,&quot; path filters.&quot;,null,null,null,null,null,null,null,null,null,[76,117,76,130]]],null,null,null,[76,31,76,131]],[25,null,null,null,null,null,null,null,null,null,[[12,null,null,null,&quot;Very simple to configure in standard &quot;,null,null,null,null,null,null,null,null,null,[76,134,76,170]],[15,null,null,null,&quot;on.pull_request.paths&quot;,null,null,null,null,null,null,null,null,null,[76,172,76,192]],[12,null,null,null,&quot;.&quot;,null,null,null,null,null,null,null,null,null,[76,194,76,194]]],null,null,null,[76,133,76,195]],[25,null,null,null,null,null,null,null,null,null,[[12,null,null,null,&quot;Touching root files runs &quot;,null,null,null,null,null,null,null,null,null,[76,198,76,222]],[19,null,null,null,null,null,null,null,null,null,[[12,null,null,null,&quot;all&quot;,null,null,null,null,null,null,null,null,null,[76,225,76,227]]],null,null,null,[76,223,76,229]],[12,null,null,null,&quot; tests, even if unrelated.&quot;,null,null,null,null,null,null,null,null,null,[76,230,76,255]]],null,null,null,[76,197,76,256]]],null,null,null,[76,1,76,257],null,null,null,null,null,[0]],[26,null,null,null,null,null,null,null,null,null,[[25,null,null,null,null,null,null,null,null,null,[[19,null,null,null,null,null,null,null,null,null,[[12,null,null,null,&quot;Dummy / Stub Workflows&quot;,null,null,null,null,null,null,null,null,null,[77,5,77,26]]],null,null,null,[77,3,77,28]]],null,null,null,[77,2,77,29]],[25,null,null,null,null,null,null,null,null,null,[[12,null,null,null,&quot;Create duplicate fallback workflows that run when &quot;,null,null,null,null,null,null,null,null,null,[77,32,77,81]],[15,null,null,null,&quot;paths-ignore&quot;,null,null,null,null,null,null,null,null,null,[77,83,77,94]],[12,null,null,null,&quot; matches to send an empty &quot;,null,null,null,null,null,null,null,null,null,[77,96,77,121]],[15,null,null,null,&quot;success&quot;,null,null,null,null,null,null,null,null,null,[77,123,77,129]],[12,null,null,null,&quot; check.&quot;,null,null,null,null,null,null,null,null,null,[77,131,77,137]]],null,null,null,[77,31,77,138]],[25,null,null,null,null,null,null,null,null,null,[[12,null,null,null,&quot;Retains per-job checks in branch protection.&quot;,null,null,null,null,null,null,null,null,null,[77,141,77,184]]],null,null,null,[77,140,77,185]],[25,null,null,null,null,null,null,null,null,null,[[12,null,null,null,&quot;Verbose and difficult to maintain across many workflows.&quot;,null,null,null,null,null,null,null,null,null,[77,188,77,243]]],null,null,null,[77,187,77,244]]],null,null,null,[77,1,77,245],null,null,null,null,null,[0]]],null,null,null,[73,1,77,245],null,null,null,null,[[0,0,0,0]]]" spk2="" dark-theme="" next-gen="" luminous="" jslog="340197;track:impression;xFZKLc:W251bGwsInVjcy1tYXJrZG93bi10YWJsZSIsbnVsbCxudWxsLG51bGwsbnVsbCxudWxsLG51bGwsbnVsbCxudWxsLDIyXQ==" class="hide-table-actions">
Approach | How it works | Pros | Cons
-- | -- | -- | --
Aggregator Job Pattern (Recommended) | Uses paths-filter + a single required ci-check job evaluated with if: always(). | Scalable, clean GitHub branch protection setup, tests execute only when needed. | Requires defining the aggregator job.
Root Path Invalidation | Add root config paths (e.g., package.json, firebase.json) to both api and web path filters. | Very simple to configure in standard on.pull_request.paths. | Touching root files runs all tests, even if unrelated.
Dummy / Stub Workflows | Create duplicate fallback workflows that run when paths-ignore matches to send an empty success check. | Retains per-job checks in branch protection. | Verbose and difficult to maintain across many workflows.

</ucs-markdown-table><hr><h3><span data-markdown-start-index="2891">Implementation Steps</span></h3><ol><li><p><strong><span data-markdown-start-index="2917">Remove top-level </span><code class="inline-code"><span data-markdown-start-index="2935">paths:</span></code></strong><span data-markdown-start-index="2944"> from the </span><code class="inline-code"><span data-markdown-start-index="2955">on.pull_request</span></code><span data-markdown-start-index="2971"> trigger so the workflow always initializes.</span></p></li><li><p><strong><span data-markdown-start-index="3020">Add a path filter step</span></strong><span data-markdown-start-index="3044"> (</span><code class="inline-code"><span data-markdown-start-index="3047">dorny/paths-filter</span></code><span data-markdown-start-index="3066">) to detect changes in </span><code class="inline-code"><span data-markdown-start-index="3090">src/api</span></code><span data-markdown-start-index="3098">, </span><code class="inline-code"><span data-markdown-start-index="3101">src/web</span></code><span data-markdown-start-index="3109">, or root configurations.</span></p></li><li><p><strong><span data-markdown-start-index="3139">Add an aggregator job</span></strong><span data-markdown-start-index="3162"> (e.g., </span><code class="inline-code"><span data-markdown-start-index="3171">ci-check</span></code><span data-markdown-start-index="3180">) that verifies </span><code class="inline-code"><span data-markdown-start-index="3197">needs.*.result</span></code><span data-markdown-start-index="3212"> does not contain </span><code class="inline-code"><span data-markdown-start-index="3231">failure</span></code><span data-markdown-start-index="3239"> or </span><code class="inline-code"><span data-markdown-start-index="3244">cancelled</span></code><span data-markdown-start-index="3254">.</span></p></li><li><p><strong><span data-markdown-start-index="3260">Update GitHub Branch Protection</span></strong><span data-markdown-start-index="3293">: In repo settings (</span><em><span data-markdown-start-index="3314">Settings &gt; Branches &gt; Branch protection rules</span></em><span data-markdown-start-index="3360">), require only </span><code class="inline-code"><span data-markdown-start-index="3377">ci-check</span></code><span data-markdown-start-index="3386"> instead of individual test job names.</span></p></li></ol></div></body></html>

### PR DESCRIPTION
**Pull request by Laurianne-M**
closes #156

### CODE REVIEW

File: .github/workflows/ci.yml [hunk 1]
```diff
@@ -0,0 +1,70 @@
...
+      web: ${{ steps.filter.outputs.web }}
+    steps:
+      - uses: actions/checkout@v4
+      - uses: dorny/paths-filter@v3
+        id: filter
+        with:
+          filters: |
+            api:
+              - 'src/api/**'
+              - 'package*.json'
```
Comment (jabaridash) [inline .github/workflows/ci.yml]: I think that this can be ignored. There is no global `package.json`, they are already inside `src/api` and `src/web`

File: .github/workflows/ci.yml
Comment (jabaridash) [inline .github/workflows/ci.yml:1]: This you be split up between the two existing files. Think about it like this. For **convenience**, I suggested that you put api and web in the same github repo. but in real life, this would be separate repos and ALL the infrastructure would be duplicated. They would be repos managed by two different teams at potentially two different parts of the planet. So whenever you need something for web, you need to redo it completely for api, and vice versa.

File: .github/workflows/ci.yml [hunk 2]
```diff
@@ -0,0 +1,70 @@
...
+          npm test --prefix src/web
+
+  # Set this single job as your required status check in GitHub branch rules
+  ci-check:
+    runs-on: ubuntu-latest
+    needs: [changes, test-api, test-web]
+    if: always()
+    steps:
+      - name: Evaluate Test Statuses
+        run: |
```
Comment (jabaridash) [inline .github/workflows/ci.yml]: We should turn this into a script that takes an argument in like `web` and `api` so that is can be reused, since we will need this logic in both `push-api.yml` and `push-web.yml`.

File: .github/workflows/push-api.yml [hunk 3]
```diff
@@ -23,7 +19,6 @@ jobs:
       run: make -C src/api test
   deploy:
     runs-on: ubuntu-24.04
-    if: github.event_name == 'push' && github.ref == 'refs/heads/main'
```
Comment (jabaridash) [inline .github/workflows/push-api.yml:26]: We still want to keep this. This is a different filter. This filter prevents deploying when pushing to a branch that is NOT main.

File: .github/workflows/push-web.yml [hunk 4]
```diff
@@ -35,7 +30,6 @@ jobs:
 
   production-deployment:
     runs-on: ubuntu-24.04
-    if: github.event_name == 'push' && github.ref == 'refs/heads/main'
```
Comment (jabaridash) [inline .github/workflows/push-web.yml:38]: We still want to keep this. This is a different filter. This filter prevents deploying when pushing to a branch that is NOT main.

File: .github/workflows/ci.yml [hunk 5]
```diff
@@ -0,0 +1,70 @@
+name: CI
+
+on:
+  pull_request:
+    branches: [main]
+
+jobs:
+  changes:
+    runs-on: ubuntu-latest
```
Comment (jabaridash) [inline .github/workflows/ci.yml]: This should be `ubuntu-24.04`. At I mentioned, for stability purposes, we should specify the version of linux we run on. That way, if there's an update, this does not break us. by accident.

File: .github/workflows/push-api.yml [hunk 6]
```diff
@@ -1,15 +1,27 @@
...
-  push:
-    branches: [ main ]
-    paths:
-      - 'src/api/**'
   pull_request:
     branches: [ main ]
-    paths:
-      - 'src/api/**'
+  push:
+    branches: [ main ]
```
Comment (jabaridash) [inline .github/workflows/push-api.yml:5]: Now that I think about it, we want this to run on every `git push` for every branch. BUT, we want to STOP certain events, like deployment. But otherwise, the tests should run for every push to every branch.

File: .github/workflows/push-web.yml [hunk 7]
```diff
@@ -33,10 +43,25 @@ jobs:
...
       run: make test-unit
 
+  # Required status check job for Web branch protection
+  web-check:
+    runs-on: ubuntu-24.04
+    needs: [changes, test]
+    if: always()
+    steps:
+      - name: Evaluate Status
+        run: |
```
Comment (jabaridash) [inline .github/workflows/push-web.yml]: This should be written as a script and used by both files. you are duplicating your work. Imagine is you wanted to change the way this works, you will have to keep doing it twice.

Comment (Laurianne-M) [reply .github/workflows/push-web.yml]: Still need to work on that, i switched the PR as a draft

Comment (jabaridash) [reply .github/workflows/push-web.yml]: Ah okay, sorry I just got a notification, that's why. Did a good thing by separating the commits.

**Comment by Laurianne-M** [comment]
Ready for review please

File: .github/scripts/evaluate-status.sh
Comment (jabaridash) [inline .github/scripts/evaluate-status.sh:1]: The name of this file is ambiguous. Evaluate the status of what?

File: .github/scripts/evaluate-status.sh [hunk 8]
```diff
@@ -0,0 +1,11 @@
+#!/usr/bin/env bash
+set -euo pipefail
+
+STATUSES="$1"
+
+if echo "$STATUSES" | grep -E "failure|cancelled"; then
```
Comment (jabaridash) [inline .github/scripts/evaluate-status.sh]: I'd put this in a function called `main()` and call it at the bottom. In the event that this script gets longer, we will already have structure.

Comment (Laurianne-M) [reply .github/scripts/evaluate-status.sh]: Done you can check my last commit

File: .github/scripts/evaluate-status.sh
Comment (Laurianne-M) [reply .github/scripts/evaluate-status.sh:1]: Done you can check my last commit

File: .github/workflows/push-api.yml [hunk 9]
```diff
@@ -1,15 +1,27 @@
 name: API Backend CI
+
```
Comment (jabaridash) [inline .github/workflows/push-api.yml:2]:
```
on: [push, pull_request]
```

File: .github/workflows/push-web.yml [hunk 10]
```diff
@@ -33,10 +43,20 @@ jobs:
     - name: Unit tests
       run: make test-unit
 
+  # Required status check job for Web branch protection
+  web-check:
```
Comment (jabaridash) [inline .github/workflows/push-web.yml]: This should be a more clear name

File: .github/workflows/push-web.yml [hunk 11]
```diff
@@ -33,10 +43,20 @@ jobs:
...
+    runs-on: ubuntu-24.04
+    needs: [changes, test]
```
Comment (jabaridash) [inline .github/workflows/push-web.yml]: I think this can depend just on `[test]`

File: .github/workflows/push-api.yml (hunk 9 above)
Comment (Laurianne-M) [reply .github/workflows/push-api.yml:2]: I did that, but it doubled my tests run as you can see bellow, I don't think that it is efficient like that? execution time still short but what if you have way more tests

Comment (jabaridash) [reply .github/workflows/push-api.yml:2]: Ok, we can delete it. Thank you for fixing it.

**Review by jabaridash — APPROVED** [review]
